### PR TITLE
Move central management settings to beat config file

### PR DESCRIPTION
**Pull request by exekias**
This change moves central management settings to the beat config, instead of having its own internal file.

In order to make this work well, `enroll` command will override existing settings to:

1. Put its own settings in the file.
2. Remove existing settings for centrally managed configurations, as they won't be allowed any longer

Before overriding the existing config, the `enroll` command will ask for confirmation and do a backup of it in all cases.

### CODE REVIEW

File: libbeat/kibana/client_config.go [hunk 1]
```diff
@@ -25,14 +25,14 @@ import (
...
 type ClientConfig struct {
-	Protocol      string            `config:"protocol"`
-	Host          string            `config:"host"`
-	Path          string            `config:"path"`
-	SpaceID       string            `config:"space.id"`
-	Username      string            `config:"username"`
-	Password      string            `config:"password"`
-	TLS           *tlscommon.Config `config:"ssl"`
-	Timeout       time.Duration     `config:"timeout"`
+	Protocol      string            `config:"protocol" yaml:"protocol"`
```
Comment (ruflin) [inline libbeat/kibana/client_config.go]: Is there a reason you changed this to yaml?

File: libbeat/cmd/instance/beat.go [hunk 2]
```diff
@@ -118,6 +118,9 @@ type beatConfig struct {
 	Pipeline   pipeline.Config `config:",inline"`
 	Monitoring *common.Config  `config:"xpack.monitoring"`
 
+	// central managmenet settings
+	Management *common.Config `config:"xpack.management"`
```
Comment (ruflin) [inline libbeat/cmd/instance/beat.go]: Should we really prefix the `management` setting by `xpack`. What about just making it `management`? I think long term it should also disappear from monitoring.

File: libbeat/common/transport/tlscommon/config.go [hunk 3]
```diff
@@ -25,14 +25,14 @@ import (
...
 type Config struct {
-	Enabled          *bool                   `config:"enabled"`
-	VerificationMode TLSVerificationMode     `config:"verification_mode"` // one of 'none', 'full'
-	Versions         []TLSVersion            `config:"supported_protocols"`
-	CipherSuites     []tlsCipherSuite        `config:"cipher_suites"`
-	CAs              []string                `config:"certificate_authorities"`
-	Certificate      CertificateConfig       `config:",inline"`
-	CurveTypes       []tlsCurveType          `config:"curve_types"`
-	Renegotiation    tlsRenegotiationSupport `config:"renegotiation"`
+	Enabled          *bool                   `config:"enabled" yml:"enabled"`
```
Comment (ruflin) [inline libbeat/common/transport/tlscommon/config.go]: Why is that needed? Some inconsistency here with yml and yaml.

File: x-pack/libbeat/cmd/enroll.go [hunk 4]
```diff
@@ -75,7 +79,7 @@ func genEnrollCmd(name, version string) *cobra.Command {
 				}
 				enrollmentToken, err = client.CreateEnrollmentToken()
 				if err != nil {
-					return errors.Wrap(err, "Creating a new enrollment token")
+					return errors.Wrap(err, "Error creating a new enrollment token")
```
Comment (ruflin) [inline x-pack/libbeat/cmd/enroll.go]: should be lower case?

File: x-pack/libbeat/management/cache.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+func (c *Cache) Save() error {
+	path := paths.Resolve(paths.Data, "management.yml")
+
+	data, err := yaml.Marshal(c)
+	if err != nil {
+		return err
+	}
+
+	// write temporary file first
+	tempFile := path + ".new"
```
Comment (ruflin) [inline x-pack/libbeat/management/cache.go]: Would be nice if we could reuse here the mechanism we use for registry writing in FB and WB

File: x-pack/libbeat/management/cache.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+	f, err := os.OpenFile(tempFile, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0600)
+	if err != nil {
+		return errors.Wrap(err, "failed to store central management settings")
+	}
+
+	_, err = f.Write(data)
+	f.Close()
```
Comment (ruflin) [inline x-pack/libbeat/management/cache.go]: What about using `defer f.Close()` instead directly after the opening? Like this if more code is added later, it will still be guaranteed that it's closed.

File: x-pack/libbeat/management/config.go [hunk 7]
```diff
@@ -5,32 +5,78 @@
...
+# have the Elasticsearch output configured, you can simply uncomment the
+# following line.
+#xpack.monitoring.elasticsearch:
+`
+
 // Config for central management
 type Config struct {
 	// true when enrolled
-	Enabled bool
+	Enabled bool `config:"enabled" yaml:"enabled"`
```
Comment (ruflin) [inline x-pack/libbeat/management/config.go]: So the yaml definition here is used to export the configs? @urso Is there an alternative to double declaration here?

File: libbeat/cmd/instance/beat.go (hunk 2 above)
Comment (exekias) [reply libbeat/cmd/instance/beat.go]: This a good point, I'm comfortable with changing this to just `management`, I just followed the same path as `xpack.monitoring`

File: x-pack/libbeat/cmd/enroll.go (hunk 4 above)
Comment (exekias) [reply x-pack/libbeat/cmd/enroll.go]: These are shown to the user (command line) so I think it's better to keep them upper case

File: x-pack/libbeat/management/cache.go (hunk 6 above)
Comment (jsoriano) [reply x-pack/libbeat/management/cache.go]: Or use `ioutil.WriteFile` instead of open, write and close?

File: x-pack/libbeat/management/cache.go (hunk 5 above)
Comment (exekias) [reply x-pack/libbeat/management/cache.go]: I think I use that already, with `file.SafeFileRotate`

File: x-pack/libbeat/management/config.go (hunk 7 above)
Comment (exekias) [reply x-pack/libbeat/management/config.go]: Yes, we use this to write management settings to the config file

File: libbeat/cmd/instance/beat.go (hunk 2 above)
Comment (exekias) [reply libbeat/cmd/instance/beat.go]: I'll change it to `management`, please shout out if you are against it

File: x-pack/libbeat/management/config.go [hunk 8]
```diff
@@ -5,32 +5,78 @@
...
+# "publish", "service".
+#logging.selectors: ["*"]
+
+#============================== Xpack Monitoring ===============================
+# {{.BeatName}} can export internal metrics to a central Elasticsearch monitoring
+# cluster.  This requires xpack monitoring to be enabled in Elasticsearch.  The
+# reporting is disabled by default.
+
+# Set to true to enable the monitoring reporter.
+#xpack.monitoring.enabled: false
```
[the remark this answers is not in the document]
Comment (exekias) [reply x-pack/libbeat/management/config.go]: I'm not making assumptions on the output settings, I hope that this is one of the first things we convert to central management, in most cases (elasticsearch output) this can be just enabled with it

**Comment by exekias** [comment]
Thanks @ruflin, I applied the requested changes
